src/compiler/checker.ts

```diff
@@ -35686,7 +35686,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     error(errorNode, Diagnostics.A_function_whose_declared_type_is_neither_void_nor_any_must_return_a_value);
                 }
             }
-            else if (type && strictNullChecks) {
+            else if (type && strictNullChecks && !isTypeAssignableTo(undefinedType, type)) {

```

Comment:
Do we want to go with this, or do we want to go with expanding the checks above?

```ts
- if (type && maybeTypeOfKind(type, TypeFlags.Undefined | TypeFlags.Void | TypeFlags.Any)) {
+ if (type && maybeTypeOfKind(type, TypeFlags.Undefined | TypeFlags.Void | TypeFlags.Any | TypeFlags.Unknown)) {
```

Reply:
The check would be faster slightly, and it allows all code to perform an earlier bail-out. I don't have a strong opinion, so let's see if Jake has a gut feel.

Reply:
I'll wait for @jakebailey's final blessing then.

Reply:
Maybe both, so that this is handled too?

```ts
type MyUnknown = unknown

function test(): MyUnknown {
	if (Math.random() > 0.5) return true;
}
```

Reply:
I am assuming the above doesn't work before either and that we need the assignability check to cover it.

I do like sticking it up at the top at least for perf, but having both feels best to me.

Reply:
If so, that sounds good, and then the code can move up (I prefer that). Deserves a test at least.

Reply:
Yes, if it works.

Reply:
For clarity - we're talking about

https://github.com/microsoft/TypeScript/pull/53490/files#diff-d9ab6589e714c71e657f601cf30ff51dfc607fc98419bf72e04f6b0fa92cc4b8R35663